<a href="https://colab.research.google.com/github/muhamadrefo/ImageClassificationColab/blob/master/SVMClassifier.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Prepare Environment

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Go to this URL in a browser: https://accounts.google.com/o/oauth2/auth?client_id=947318989803-6bn6qk8qdgf4n4g3pfee6491hc0brc4i.apps.googleusercontent.com&redirect_uri=urn%3aietf%3awg%3aoauth%3a2.0%3aoob&response_type=code&scope=email%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdocs.test%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive%20https%3a%2f%2fwww.googleapis.com%2fauth%2fdrive.photos.readonly%20https%3a%2f%2fwww.googleapis.com%2fauth%2fpeopleapi.readonly

Enter your authorization code:
··········
Mounted at /content/drive


In [3]:
#Importing Packages

import numpy as np
import matplotlib.pyplot as plt
import csv
from sklearn import svm
from sklearn.metrics import accuracy_score
from sklearn.metrics import confusion_matrix as cm
from sklearn import model_selection
from sklearn.model_selection import train_test_split as tt
from sklearn.metrics import classification_report

In [2]:
#filtered dataset
%cd /content/drive/My Drive/TestingPlace/PlaceofDatabase/CNN/

#Unfilter dataset
#%cd /content/drive/My Drive/Batik/

/content/drive/My Drive/TestingPlace/PlaceofDatabase/CNN


# Reading the Dataset

In [4]:
import pandas as pd
#data = pd.read_csv('DataValidation.csv')
data = pd.read_csv('DataTraining.csv')

#Encoding the Categorical Data

In [5]:
from sklearn.preprocessing import LabelEncoder
le = LabelEncoder()
df = data.apply(le.fit_transform)
df.head(15)

Path  Dissimilarity  Contrast  Homogeneity  Energy  Correlation  ASM  Label
0     20            299       273          104     118           65  118      0
1     17            243       195          154     153           97  153      0
2     18            277       247          123     149           60  149      0
3     19            288       244          107     159           50  159      0
4     16            285       250          114     120           98  120      0
5     14            309       294           97     119           59  119      0
6     15            256       217          137     175           67  175      0
7      9            310       293           91     134           35  134      0
8     10            279       245          118     127           73  127      0
9     12            249       201          144     163           94  163      0
10    11            252       216          145     151           93  151      0
11    13            211       189          194     285          201  285      0
12     8            280       252          120     138           64  138      0
13     7            293       259          106     141           57  141      0
14     6            220       199          182     317           38  317      0

#Splitting the Data into Training and Testing Data

In [6]:
X = df.iloc[:, 1:6].values
Y = df.iloc[:, 7].values
X_train, X_test, Y_train, Y_test = tt(X,Y, test_size=0.2, random_state=4)

In [7]:
X_train

array([[357, 359,  58,  47, 259],
       [106,  85, 302, 231, 246],
       [232, 242, 188,  72, 399],
       ...,
       [359, 356,  56, 167,   8],
       [241, 246, 176, 188, 179],
       [265, 258, 143, 125, 128]])

In [8]:
X_train.shape

(340, 5)

#Model SVM Start Here

**Polynomial**

In [9]:
# Kernel Polynomial
modelSVM = svm.SVC(C=1.0, kernel='poly', gamma='scale', probability=True)
modelSVM.fit(X_train, Y_train)
y_pred = modelSVM.predict(X_test)

In [10]:
print("histogram y_pred\n",np.histogram(y_pred))
print("histogram Y_test\n",np.histogram(Y_test))
print("Confusion Matriks\n", cm(Y_test, modelSVM.predict(X_test)))

histogram y_pred
 (array([45,  0,  0,  0,  0,  0,  0,  0,  0, 40]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
histogram Y_test
 (array([49,  0,  0,  0,  0,  0,  0,  0,  0, 36]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
Confusion Matriks
 [[43  6]
 [ 2 34]]


In [11]:
print("Classification Report:\n", classification_report(Y_test,y_pred))

Classification Report:
               precision    recall  f1-score   support

           0       0.96      0.88      0.91        49
           1       0.85      0.94      0.89        36

    accuracy                           0.91        85
   macro avg       0.90      0.91      0.90        85
weighted avg       0.91      0.91      0.91        85



In [12]:
print("Accuracy Score: ", accuracy_score(np.delete(Y_test, np.where(np.isnan(X_test).any(axis=1))), y_pred))

Accuracy Score:  0.9058823529411765


**RBF**

In [13]:
modelSVM = svm.SVC(C=1.0, kernel='rbf', gamma='scale', probability=True)
modelSVM.fit(X_train, Y_train)
y_pred = modelSVM.predict(X_test)

print("histogram y_pred\n",np.histogram(y_pred))
print("histogram Y_test\n",np.histogram(Y_test))
print("Confusion Matriks\n", cm(Y_test, modelSVM.predict(X_test)))

histogram y_pred
 (array([42,  0,  0,  0,  0,  0,  0,  0,  0, 43]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
histogram Y_test
 (array([49,  0,  0,  0,  0,  0,  0,  0,  0, 36]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
Confusion Matriks
 [[40  9]
 [ 2 34]]


In [14]:
print("Classification Report:\n", classification_report(Y_test,y_pred))
print("Accuracy Score: ", accuracy_score(np.delete(Y_test, np.where(np.isnan(X_test).any(axis=1))), y_pred))

Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.82      0.88        49
           1       0.79      0.94      0.86        36

    accuracy                           0.87        85
   macro avg       0.87      0.88      0.87        85
weighted avg       0.88      0.87      0.87        85

Accuracy Score:  0.8705882352941177


**SIGMOID**

In [15]:
modelSVM = svm.SVC(C=1.0, kernel='sigmoid', gamma='scale', probability=True)
modelSVM.fit(X_train, Y_train)
y_pred = modelSVM.predict(X_test)

print("histogram y_pred\n",np.histogram(y_pred))
print("histogram Y_test\n",np.histogram(Y_test))
print("Confusion Matriks\n", cm(Y_test, modelSVM.predict(X_test)))

histogram y_pred
 (array([37,  0,  0,  0,  0,  0,  0,  0,  0, 48]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
histogram Y_test
 (array([49,  0,  0,  0,  0,  0,  0,  0,  0, 36]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
Confusion Matriks
 [[14 35]
 [23 13]]


In [16]:
print("Classification Report:\n", classification_report(Y_test,y_pred))
print("Accuracy Score: ", accuracy_score(np.delete(Y_test, np.where(np.isnan(X_test).any(axis=1))), y_pred))

Classification Report:
               precision    recall  f1-score   support

           0       0.38      0.29      0.33        49
           1       0.27      0.36      0.31        36

    accuracy                           0.32        85
   macro avg       0.32      0.32      0.32        85
weighted avg       0.33      0.32      0.32        85

Accuracy Score:  0.3176470588235294


**LINEAR**

In [17]:
modelSVM = svm.SVC(C=1.0, kernel='linear', gamma='scale', probability=True)
modelSVM.fit(X_train, Y_train)
y_pred = modelSVM.predict(X_test)

print("histogram y_pred\n",np.histogram(y_pred))
print("histogram Y_test\n",np.histogram(Y_test))
print("Confusion Matriks\n", cm(Y_test, modelSVM.predict(X_test)))

histogram y_pred
 (array([53,  0,  0,  0,  0,  0,  0,  0,  0, 32]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
histogram Y_test
 (array([49,  0,  0,  0,  0,  0,  0,  0,  0, 36]), array([0. , 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1. ]))
Confusion Matriks
 [[36 13]
 [17 19]]


In [18]:
print("Classification Report:\n", classification_report(Y_test,y_pred))
print("Accuracy Score: ", accuracy_score(np.delete(Y_test, np.where(np.isnan(X_test).any(axis=1))), y_pred))

Classification Report:
               precision    recall  f1-score   support

           0       0.68      0.73      0.71        49
           1       0.59      0.53      0.56        36

    accuracy                           0.65        85
   macro avg       0.64      0.63      0.63        85
weighted avg       0.64      0.65      0.64        85

Accuracy Score:  0.6470588235294118


#Analytic

Using filtered Dataset
Accuracy Score Linear: 0.64
Accuracy Score Sigmoid:0.31
Accuracy Score RBF: 0.87
Accuracy Score Polynomial: 0.90

Using Unfiltered Dataset
Accuracy Score Linear: 0.62
Accuracy Score Sigmoid:0.30
Accuracy Score RBF: 0.86
Accuracy Score Polynomial: 0.89